api/python/cellxgene_census/src/cellxgene_census/experimental/_embedding.py

```diff
@@ -0,0 +1,113 @@
+        obs_soma_joinids:
+            The slice of the embedding to fetch and return.
+        context:
+            A custom :class:`SOMATileDBContext` which will be used to open the SOMA object.
+            Optional, defaults to None.
+
+    Returns:
+        A :class:`numpy.ndarray` containing the embeddings. Embeddings are positionally
+        indexed by the obs_soma_joinids. In other words, the cell identified by
+        `obs_soma_joinids[i]` corresponds to the `ith` positoin in the returned ndarray.
+
+    Raises:
+        ValueError: if the Census and embedding are mismatched.
+
+    Lifecycle:
+        experimental
+
+    Examples:
+        >>> obs_somaids_to_fetch = np.array([10,11], dtype=np.int64)
+        >>> emb = cellxgene_census.experimental.get_embedding('2023-10-23', embedding_uri, obs_somaids_to_fetch)
+        >>> emb.shape
+        (2, 200)
+        >>> emb[:, 0:4]
+        array([[ 0.02954102,  1.0390625 , -0.14550781, -0.40820312],
+            [-0.00224304,  1.265625  ,  0.05883789, -0.7890625 ]],
+            dtype=float32)
+
+    """
+
+    if isinstance(obs_soma_joinids, (pa.Array, pa.ChunkedArray)):
+        obs_soma_joinids = obs_soma_joinids.to_numpy()
+    if obs_soma_joinids.dtype != np.int64:
+        raise TypeError("obs_soma_joinids must be array of int64")
+
+    # Currently, all embeddings are hosted in us-west-2 so use that as a default.
+    # Allow the user to override for exceptional cases.
+    context = _build_soma_tiledb_context("us-west-2", context)
+
+    # Attempt to resolve census version aliases
+    census_directory = get_census_version_directory()
+    resolved_census_version = census_directory.get(census_version, None)
+
+    with soma.open(embedding_uri, context=context) as E:
+        embedding_metadata = json.loads(E.metadata["CxG_contrib_metadata"])
+
+        if resolved_census_version is None:
+            warnings.warn(
+                "Unable to determine Census version - skipping validation of Census and embedding version.",
+                stacklevel=1,
+            )
+        elif resolved_census_version != census_directory.get(embedding_metadata["census_version"], None):
+            raise ValueError("Census and embedding mismatch - census_version not equal")
+
+        with open_soma(census_version=census_version, context=context) as census:
+            experiment_name = embedding_metadata["experiment_name"]
+            if experiment_name not in census["census_data"]:
+                raise ValueError("Census and embedding mismatch - experiment_name does not exist")
+            measurement_name = embedding_metadata["measurement_name"]
+            if measurement_name not in census["census_data"][experiment_name].ms:
+                raise ValueError("Census and embedding mismatch - measurement_name does not exist")
+
+        embedding_shape = (len(obs_soma_joinids), E.shape[1])
+        embedding = np.full(embedding_shape, np.NaN, dtype=np.float32, order="C")
```

Comment:
Based on some experiments I did while implementing the `to_anndata` in SOMA, embeddings have some values that are exactly zero (_very_ few - I guess those are very small values that underflow). This will cause them to be NaN instead, which I believe it's not what we want.

Reply:
The embeddings are stored in sparse format, so zeros correspond to missing data. Some models were not trained on certain cells. For example, most models were only trained on primary data.

Reply:
Actually, this will cause them to be zero, not NaN, as we _explicitly store zero values for all embeddings.  The only missing values are "no embedding", not "zero-valued embedding".

See the spec [here](https://github.com/chanzuckerberg/cellxgene-census/blob/bkmartinjr/contrib-embeddings/tools/census_contrib/embedding_metadata.md), in particular, the paragraph labelled *important*

Reply:
Yeah, I realized that my experiments were run against an initial prototype where the zeros were not explicitly marked. Good to know that the final version won't have this issue.